src/os/bluestore/BlueFS.h

```diff
@@ -235,6 +235,38 @@ class BlueFS {
   struct File : public RefCountedObject {
     MEMPOOL_CLASS_HELPERS();
 
+    struct WALFlush {
```

Comment:
It would be nice to have a comprehensive comment on how WALv2 envelope looks.

I assume that it works like:
`1st: LENGTHdatadatadataMARKER000000---------------------------------------------------------`
`2nd: xxxxxxxxxxxxxxxxxxxxxxxxLENGTHdatadatadataMARKER000000---------------------------------`
LENGTH of next envelope overwrites pre-written 000000.
MARKER is file's ino.

I do not think that appending and later overwriting "0000" is worth it.
Benefit:
One can read next-length, see it is "0" and conclude WAL is done.
Cost:
FileWriter must go back and overwrite what was already written.
Modification of flush_data is needed.
Problem case:
`LENGTHdatadatadataMARKER000000`
`wwwwwwwwwwwwwwwwwwwwwwwwwwwfff`
w = data written f=data not written, power out
when reading we get 
`LENGTHdatadatadataMARKER000gggggggggggggggggggggggggggggggg`
and we decide to interpret garbage g as proper data.

Alternative solution: do not write "000000" and relay on MARKER.
Benefit:
Can reuse flush_data as-is.
Cost:
Must read LENGTH, read length data and verify that MARKER has proper value.
Problem case:
`LENGTHdatadatadataMARKER`
`wwwwwwwwwwwwwwwwwwwwwfff`
when reading we get corruption on MARKER and conclude there is no entry.

Reply:
I do not think MARKER should be something as simple as ino.
I would imagine that MARKER would be some random or a hash(ino) to get high bit entropy.

Reply:
> I do not think that appending and later overwriting "0000" is worth it. Benefit: One can read next-length, see it is "0" and conclude WAL is done. Cost: FileWriter must go back and overwrite what was already written. Modification of flush_data is needed. Problem case: `LENGTHdatadatadataMARKER000000` `wwwwwwwwwwwwwwwwwwwwwwwwwwwfff` w = data written f=data not written, power out when reading we get `LENGTHdatadatadataMARKER000gggggggggggggggggggggggggggggggg` and we decide to interpret garbage g as proper data.

yes, we treat g as proper data, but marker will most like fail and therefore produce a corruption error which is good.

Is this statement, "Cost: FileWriter must go back and overwrite what was already written", correct? If you are overwriting it doesn't cost anything. The write if there were 0s or not would be on the same place with the same data. The only cost I would say is adding 8 bytes to the disk write.

> Alternative solution: do not write "000000" and relay on MARKER. Benefit: Can reuse flush_data as-is. Cost: Must read LENGTH, read length data and verify that MARKER has proper value. Problem case: `LENGTHdatadatadataMARKER` `wwwwwwwwwwwwwwwwwwwwwfff` when reading we get corruption on MARKER and conclude there is no entry.

Looks good on hand but this means I must rely on `fnode.size` to deal with **ambigous** `bad(marker)` (assuming bad is a function that determines marker is not correct) because if it is bad it could mean two things: EOF or corruption. To determine whether which of the two is the case you need fnode.size to be updated with proper log transactions but that defeats the purpose of new wal reducing fdatasyncs?

Reply:
Overwriting "000000" for me has the main cost of reduced simplicity.
Without it, we can imagine data conversion process as:
rocks input -> WAL envelope -> original write()

Plus, future extension: If we decide to use zoned devices (we cannot overwrite, only expand) for WALv2 we should be able to handle sequence:
`data0data0data0------------` `data1data1-----------------` `data2data2data2data2----`
where "-" is empty, unused part of blocks.


`fnode.size` should always be in sync with some marker.
I imagine that we capture `fnode.size` of a previous wal envelope, one that is already stable on disk.
We do not force flush_and_sync_log, but we can update `fnode.size` and set "dirty", so if some flush_and_sync_log happens, it will catch it.
That way, we could relay on the value to be a stable end of envelope, but maybe not the last.

Reply:
from offline conversation this is already implemented in new commits.